spec/lib/link_checker.rb

```diff
@@ -0,0 +1,165 @@
+            "Nothing hidden here",
+            { "anotherHash" => "this string here is another http://www.banana.com in this string" },
+          ],
+          { "secondHash" => "the second member also has an url https://extrahelp.co.uk woah",
+            "thirdHash" => "the second member also has an url http://terror.com", "false" => false }
+        ],
+      })
+      expect(result).to eq(Set[
+        "https://www.example.com",
+        "http://www.banana.com",
+        "https://extrahelp.co.uk",
+        "https://example.gov.uk",
+        "http://terror.com",
+      ])
+    end
+  end
+
+  describe "get_urls_from_locale_files method" do
+    it "should separate out gov urls, and example urls" do
+      allow(Dir).to receive(:glob).with("config/locales/*.yml") { %i[dummy_path1] }
+      allow(File).to receive(:read).with(:dummy_path1) { :dummy_file1 }
+      allow(YAML).to receive(:safe_load).with(:dummy_file1) { :dummy_yaml1 }
+      allow(LinkCheckerMethods).to receive(:gather_urls).with(:dummy_yaml1) {
+        Set["http://buildings.org", "http://defra.gov.uk", "https://www.abstract.com", "https://banana.gov.uk"]
+      }
+
+      govuk_urls, other_urls = LinkCheckerMethods.get_urls_from_locale_files
+      expect(govuk_urls).to match_array(["https://banana.gov.uk", "http://defra.gov.uk"])
+      expect(other_urls).to match_array(["http://buildings.org", "https://www.abstract.com"])
+    end
+
+    it "should iterate over all files returned by glob" do
+      allow(Dir).to receive(:glob).with("config/locales/*.yml") { %i[dummy_path1 dummy_path2] }
+      allow(File).to receive(:read).with(:dummy_path1) { :dummy_file1 }
+      allow(File).to receive(:read).with(:dummy_path2) { :dummy_file2 }
+      allow(YAML).to receive(:safe_load).with(:dummy_file1) { :dummy_yaml1 }
+      allow(YAML).to receive(:safe_load).with(:dummy_file2) { :dummy_yaml2 }
+      allow(LinkCheckerMethods).to receive(:gather_urls).with(:dummy_yaml1) {
+        Set["http://houses.gov.uk", "https://www.abstract.com"]
+      }
+      allow(LinkCheckerMethods).to receive(:gather_urls).with(:dummy_yaml2) {
+        Set["https://bins.gov.uk", "http://www.example.com"]
+      }
+
+      govuk_urls, other_urls = LinkCheckerMethods.get_urls_from_locale_files
+      expect(govuk_urls).to match_array(["https://bins.gov.uk", "http://houses.gov.uk"])
+      expect(other_urls).to match_array(["http://www.example.com", "https://www.abstract.com"])
+    end
+  end
+
+  describe "find_invalid_govuk_paths" do
+    it "should separate out gov urls, and example urls" do
+      test_paths = ["/test/path", "/another/test/path", "/a/third/test/path", "/fifth", "/sixth", "/seventh"]
+      test_urls = test_paths.map { |path| "http://www.gov.uk#{path}" }
+
+      mock_client = instance_double("GdsApi::PublishingApi")
+      expect(GdsApi::PublishingApi).to receive(:new).with("https://publishing-api.publishing.service.gov.uk", bearer_token: "test_token") {
```

Comment:
The GDS API Adapters gem comes with handy methods for testing (including stubs https://github.com/alphagov/gds-api-adapters/blob/master/lib/gds_api/test_helpers/link_checker_api.rb)

Reply:
Having taken a look at this, the stub method doesn't allow you to vary the response based on the value of `with_drafts`, which makes it not fit for our purposes.

In terms of test validity, I'm using an instance double, which should, if the method signature changes too much, result in test failures.